File: ios-app/ChainlessChain/Features/Knowledge/Services/VectorStore.swift
import Foundation
import CoreCommon
import Accelerate

/// Vector Store - Manages vector storage and similarity search
/// Reference: desktop-app-vue/src/main/vector/vector-store.js
/// iOS implementation uses SQLite-based persistent storage via VectorStoreRepository
@MainActor
class VectorStore: ObservableObject {
    private let logger = Logger.shared

    // Configuration
    private let similarityThreshold: Float
    private let topK: Int

    // Persistence repository
    private let repository = VectorStoreRepository.shared

    // In-memory cache for fast access
    private var vectors: [String: VectorEntry] = [:]
    private var isCacheLoaded = false

    @Published var isInitialized = false

    struct VectorEntry {
        let id: String
        let embedding: [Float]
        let metadata: VectorMetadata
        let document: String
        let timestamp: Date
    }

    struct VectorMetadata: Codable {
        let title: String
        let type: String
        let createdAt: String?
        let updatedAt: String?
    }

    struct SearchResult {
        let id: String
        let score: Float
        let metadata: VectorMetadata
        let document: String
    }

    init(similarityThreshold: Float = 0.6, topK: Int = 10) {
        self.similarityThreshold = similarityThreshold
        self.topK = topK
    }

    /// Initialize vector store
    func initialize() async -> Bool {
        logger.debug("[VectorStore] Initializing vector store (SQLite persistent mode)...")

        // Load vectors from repository into memory cache
        await loadFromRepository()

        isInitialized = true
        logger.debug("[VectorStore] Vector store initialized successfully with \(vectors.count) vectors")

        return true
    }

    /// Load all vectors from repository into memory cache
    private func loadFromRepository() async {
        guard !isCacheLoaded else { return }

        do {
            let entries = try repository.getAllVectors()
            vectors.removeAll()

            for entry in entries {
                let localEntry = VectorEntry(
                    id: entry.id,
                    embedding: entry.embedding,
                    metadata: VectorMetadata(
                        title: entry.metadata.title,
                        type: entry.metadata.type,
                        createdAt: entry.metadata.createdAt,
                        updatedAt: entry.metadata.updatedAt
                    ),
                    document: entry.document,
                    timestamp: entry.timestamp
                )
                vectors[entry.id] = localEntry
            }

            isCacheLoaded = true
            logger.debug("[VectorStore] Loaded \(entries.count) vectors from repository")

        } catch {
            logger.error("[VectorStore] Failed to load vectors from repository: \(error)")
        }
    }

    /// Add a single vector
    func addVector(id: String, embedding: [Float], metadata: VectorMetadata, document: String) async throws {
        let timestamp = Date()
        let entry = VectorEntry(
            id: id,
            embedding: embedding,
            metadata: metadata,
            document: document,
            timestamp: timestamp
        )

        // Persist to repository
        let repoEntry = RepoVectorEntry(
            id: id,
            embedding: embedding,
            metadata: RepoVectorMetadata(
                title: metadata.title,
                type: metadata.type,
                createdAt: metadata.createdAt,
                updatedAt: metadata.updatedAt
            ),
            document: document,
            timestamp: timestamp
        )

        try repository.saveVector(repoEntry)

        // Update memory cache
        vectors[id] = entry
        logger.debug("[VectorStore] Added vector: \(metadata.title)")
    }

    /// Add multiple vectors in batch
    func addVectorsBatch(items: [(id: String, embedding: [Float], metadata: VectorMetadata, document: String)]) async throws -> Int {
        var repoEntries: [RepoVectorEntry] = []
        let timestamp = Date()

        for item in items {
            let entry = VectorEntry(
                id: item.id,
                embedding: item.embedding,
                metadata: item.metadata,
                document: item.document,
                timestamp: timestamp
            )
            vectors[item.id] = entry

            // Prepare for batch save
            repoEntries.append(RepoVectorEntry(
                id: item.id,
                embedding: item.embedding,
                metadata: RepoVectorMetadata(
                    title: item.metadata.title,
                    type: item.metadata.type,
                    createdAt: item.metadata.createdAt,
                    updatedAt: item.metadata.updatedAt
                ),
                document: item.document,
                timestamp: timestamp
            ))
        }

        // Batch persist to repository
        try repository.saveVectorsBatch(repoEntries)

        logger.debug("[VectorStore] Batch added \(items.count) vectors")
        return items.count
    }

    /// Update a vector
    func updateVector(id: String, embedding: [Float], metadata: VectorMetadata? = nil) async throws {
        guard var entry = vectors[id] else {
            throw VectorStoreError.vectorNotFound(id)
        }

        let newMetadata = metadata ?? entry.metadata
        let timestamp = Date()

        entry = VectorEntry(
            id: entry.id,
            embedding: embedding,
            metadata: newMetadata,
            document: entry.document,
            timestamp: timestamp
        )

        // Persist to repository
        let repoEntry = RepoVectorEntry(
            id: entry.id,
            embedding: embedding,
            metadata: RepoVectorMetadata(
                title: newMetadata.title,
                type: newMetadata.type,
                createdAt: newMetadata.createdAt,
                updatedAt: newMetadata.updatedAt
            ),
            document: entry.document,
            timestamp: timestamp
        )
        try repository.saveVector(repoEntry)

        vectors[id] = entry
        logger.debug("[VectorStore] Updated vector: \(id)")
    }

    /// Delete a vector
    func deleteVector(id: String) async throws {
        guard vectors.removeValue(forKey: id) != nil else {
            throw VectorStoreError.vectorNotFound(id)
        }

        // Delete from repository
        try repository.deleteVector(id: id)

        logger.debug("[VectorStore] Deleted vector: \(id)")
    }

    /// Search for similar vectors using optimized min-heap algorithm
    /// - Parameters:
    ///   - queryEmbedding: Query vector
    ///   - topK: Number of results to return (optional, uses default if nil)
    ///   - filter: Metadata filter (optional)
    /// - Returns: Array of search results sorted by similarity (O(n log k) complexity)
    func search(queryEmbedding: [Float], topK: Int? = nil, filter: [String: String]? = nil) async throws -> [SearchResult] {
        let k = topK ?? self.topK

        // Use a min-heap to maintain top-K results efficiently
        // This gives us O(n log k) instead of O(n log n)
        var minHeap = MinHeap<ScoredResult>(capacity: k)

        // Calculate similarity for all vectors
        for (id, entry) in vectors {
            // Apply filter if provided
            if let filter = filter {
                var matches = true
                for (key, value) in filter {
                    let metadataValue: String?
                    switch key {
                    case "type":
                        metadataValue = entry.metadata.type
                    case "title":
                        metadataValue = entry.metadata.title
                    default:
                        metadataValue = nil
                    }

                    if metadataValue != value {
                        matches = false
                        break
                    }
                }

                if !matches {
                    continue
                }
            }

            let similarity = cosineSimilarity(queryEmbedding, entry.embedding)

            // Only consider results above threshold
            if similarity >= similarityThreshold {
                let scored = ScoredResult(
                    id: id,
                    score: similarity,
                    metadata: entry.metadata,
                    document: entry.document
                )

                // Add to heap, automatically maintaining top-K
                minHeap.insert(scored)
            }
        }

        // Extract results in descending order of score
        let results = minHeap.extractAllDescending().map { scored in
            SearchResult(
                id: scored.id,
                score: scored.score,
                metadata: scored.metadata,
                document: scored.document
            )
        }

        logger.debug("[VectorStore] Search returned \(results.count) results")
        return results
    }

    /// Internal scored result for heap operations
    private struct ScoredResult: Comparable {
        let id: String
        let score: Float
        let metadata: VectorMetadata
        let document: String

        // Min-heap comparison (lower score = higher priority for removal)
        static func < (lhs: ScoredResult, rhs: ScoredResult) -> Bool {
            return lhs.score < rhs.score
        }
    }

    /// Min-heap implementation for efficient top-K selection
    private struct MinHeap<T: Comparable> {
        private var heap: [T] = []
        private let capacity: Int

        init(capacity: Int) {
            self.capacity = capacity
            heap.reserveCapacity(capacity + 1)
        }

        mutating func insert(_ element: T) {
            if heap.count < capacity {
                heap.append(element)
                siftUp(heap.count - 1)
            } else if element > heap[0] {
                // New element is larger than smallest in heap
                heap[0] = element
                siftDown(0)
            }
        }

        mutating func extractAllDescending() -> [T] {
            // Sort in descending order (highest first)
            return heap.sorted { $0 > $1 }
        }

        private mutating func siftUp(_ index: Int) {
            var child = index
            var parent = (child - 1) / 2
            while child > 0 && heap[child] < heap[parent] {
                heap.swapAt(child, parent)
                child = parent
                parent = (child - 1) / 2
            }
        }

        private mutating func siftDown(_ index: Int) {
            var parent = index
            while true {
                let left = 2 * parent + 1
                let right = 2 * parent + 2
                var smallest = parent

                if left < heap.count && heap[left] < heap[smallest] {
                    smallest = left
                }
                if right < heap.count && heap[right] < heap[smallest] {
                    smallest = right
                }
                if smallest == parent {
                    break
                }
                heap.swapAt(parent, smallest)
                parent = smallest
            }
        }
    }

    /// Calculate cosine similarity between two vectors using SIMD (Accelerate framework)
    private func cosineSimilarity(_ vecA: [Float], _ vecB: [Float]) -> Float {
        guard vecA.count == vecB.count, !vecA.isEmpty else {
            return 0
        }

        let count = vDSP_Length(vecA.count)

        // Calculate dot product using SIMD
        var dotProduct: Float = 0
        vDSP_dotpr(vecA, 1, vecB, 1, &dotProduct, count)

        // Calculate squared norms using SIMD
        var normASq: Float = 0
        var normBSq: Float = 0
        vDSP_svesq(vecA, 1, &normASq, count)
        vDSP_svesq(vecB, 1, &normBSq, count)

        guard normASq > 0, normBSq > 0 else {
            return 0
        }

        return dotProduct / (sqrt(normASq) * sqrt(normBSq))
    }

    /// Get vector store statistics
    func getStats() async -> VectorStoreStats {
        do {
            let repoStats = try repository.getStatistics()
            return VectorStoreStats(
                mode: repoStats.mode,
                count: repoStats.vectorCount,
                collectionName: "knowledge_base",
                storageSizeMB: repoStats.storageSizeMB,
                cacheCount: repoStats.cacheCount
            )
        } catch {
            return VectorStoreStats(
                mode: "sqlite",
                count: vectors.count,
                collectionName: "knowledge_base",
                storageSizeMB: 0,
                cacheCount: 0
            )
        }
    }

    /// Clear all vectors
    func clear() async throws {
        // Clear repository
        try repository.deleteAllVectors()

        // Clear memory cache
        vectors.removeAll()
        isCacheLoaded = false

        logger.debug("[VectorStore] Vector store cleared")
    }

    /// Reload vectors from repository (refresh cache)
    func refresh() async {
        isCacheLoaded = false
        await loadFromRepository()
        logger.debug("[VectorStore] Vector store refreshed from repository")
    }

    /// Rebuild index from knowledge items
    /// - Parameters:
    ///   - items: Knowledge items to index
    ///   - embeddingFn: Function to generate embeddings
    func rebuildIndex(items: [KnowledgeItem], embeddingFn: (String) async throws -> [Float]) async throws {
        logger.debug("[VectorStore] Rebuilding index for \(items.count) items...")

        // Clear existing index
        try await clear()

        let batchSize = 10
        var processed = 0

        for i in stride(from: 0, to: items.count, by: batchSize) {
            let endIndex = min(i + batchSize, items.count)
            let batch = Array(items[i..<endIndex])

            // Generate embeddings for batch
            var batchItems: [(id: String, embedding: [Float], metadata: VectorMetadata, document: String)] = []

            for item in batch {
                let text = "\(item.title)\n\(item.content ?? "")"
                let embedding = try await embeddingFn(text)

                let metadata = VectorMetadata(
                    title: item.title,
                    type: item.type,
                    createdAt: item.createdAt,
                    updatedAt: item.updatedAt
                )

                batchItems.append((
                    id: item.id,
                    embedding: embedding,
                    metadata: metadata,
                    document: item.content ?? item.title
                ))
            }

            // Add batch to vector store
            _ = try await addVectorsBatch(items: batchItems)

            processed += batch.count
            let percentage = Int((Float(processed) / Float(items.count)) * 100)
            logger.debug("[VectorStore] Rebuild progress: \(processed)/\(items.count) (\(percentage)%)")
        }

        logger.debug("[VectorStore] Index rebuild complete")
    }

    struct VectorStoreStats {
        let mode: String
        let count: Int
        let collectionName: String
        var storageSizeMB: Double = 0
        var cacheCount: Int = 0
    }
}

// MARK: - Error Types

enum VectorStoreError: LocalizedError {
    case vectorNotFound(String)
    case invalidEmbedding
    case searchFailed(String)

    var errorDescription: String? {
        switch self {
        case .vectorNotFound(let id):
            return "Vector not found: \(id)"
        case .invalidEmbedding:
            return "Invalid embedding vector"
        case .searchFailed(let message):
            return "Search failed: \(message)"
        }
    }
}

File: desktop-app-vue/src/main/vector/vector-store.js
/**
 * 向量存储服务
 *
 * 使用ChromaDB进行向量存储和检索
 */

const { logger } = require("../utils/logger.js");
const { ChromaClient } = require("chromadb");
const EventEmitter = require("events");
const path = require("path");
const { app } = require("electron");
const fs = require("fs");

/**
 * 向量存储配置
 */
const DEFAULT_CONFIG = {
  // ChromaDB配置
  chromaUrl: "http://localhost:8000", // ChromaDB服务地址
  collectionName: "knowledge_base", // 集合名称

  // 向量参数
  embeddingDimension: 768, // 向量维度 (bge-small-zh-v1.5)
  distanceMetric: "cosine", // 距离度量: cosine, l2, ip

  // 检索参数
  topK: 5, // 返回Top-K结果
  similarityThreshold: 0.7, // 相似度阈值

  // 本地缓存
  enableCache: true, // 是否启用本地缓存
  cacheDir: null, // 缓存目录 (自动设置)
};

/**
 * 向量存储管理类
 */
class VectorStore extends EventEmitter {
  constructor(config = {}) {
    super();

    this.config = { ...DEFAULT_CONFIG, ...config };

    // 设置缓存目录
    if (!this.config.cacheDir) {
      const userDataPath = app.getPath("userData");
      this.config.cacheDir = path.join(userDataPath, "vector-cache");
    }

    this.client = null;
    this.collection = null;
    this.isInitialized = false;

    // 内存缓存
    this.memoryCache = new Map();
  }

  /**
   * 初始化向量存储
   */
  async initialize() {
    logger.info("[VectorStore] 初始化向量存储...");
    logger.info("[VectorStore] ChromaDB地址:", this.config.chromaUrl);

    try {
      // 确保缓存目录存在
      if (this.config.enableCache && !fs.existsSync(this.config.cacheDir)) {
        fs.mkdirSync(this.config.cacheDir, { recursive: true });
      }

      // 连接ChromaDB (使用新的host/port配置而不是废弃的path参数)
      const url = new URL(this.config.chromaUrl);
      this.client = new ChromaClient({
        host: url.hostname,
        port: url.port || (url.protocol === "https:" ? 443 : 8000),
      });

      // 检查连接
      await this.checkConnection();

      // 获取或创建集合
      await this.getOrCreateCollection();

      this.isInitialized = true;
      logger.info("[VectorStore] 向量存储初始化成功");
      this.emit("initialized");

      return true;
    } catch (error) {
      logger.error("[VectorStore] 初始化失败:", error);

      // 如果ChromaDB未运行，使用内存模式
      logger.warn("[VectorStore] 切换到内存模式");
      this.isInitialized = false;
      this.emit("fallback-memory");

      return false;
    }
  }

  /**
   * 检查ChromaDB连接
   */
  async checkConnection() {
    try {
      const version = await this.client.version();
      logger.info("[VectorStore] ChromaDB版本:", version);
      return true;
    } catch (error) {
      throw new Error("无法连接到ChromaDB服务");
    }
  }

  /**
   * 获取或创建集合
   */
  async getOrCreateCollection() {
    try {
      // 尝试获取现有集合
      this.collection = await this.client.getCollection({
        name: this.config.collectionName,
      });

      logger.info("[VectorStore] 使用现有集合:", this.config.collectionName);
    } catch (error) {
      // 集合不存在，创建新集合
      logger.info("[VectorStore] 创建新集合:", this.config.collectionName);

      this.collection = await this.client.createCollection({
        name: this.config.collectionName,
        metadata: {
          "hnsw:space": this.config.distanceMetric,
          description: "ChainlessChain知识库向量存储",
        },
      });
    }

    // 获取集合统计
    const count = await this.collection.count();
    logger.info("[VectorStore] 集合中的向量数:", count);
  }

  /**
   * 添加向量
   * @param {Object} item - 知识库项
   * @param {Array} embedding - 嵌入向量
   */
  async addVector(item, embedding) {
    if (!this.isInitialized || !this.collection) {
      // 内存模式
      this.memoryCache.set(item.id, { item, embedding });
      return;
    }

    try {
      await this.collection.add({
        ids: [item.id],
        embeddings: [embedding],
        metadatas: [
          {
            title: item.title,
            type: item.type,
            created_at: item.created_at,
            updated_at: item.updated_at,
          },
        ],
        documents: [item.content || item.title],
      });

      logger.info(`[VectorStore] 添加向量: ${item.title}`);
      this.emit("vector-added", item.id);

      return true;
    } catch (error) {
      logger.error("[VectorStore] 添加向量失败:", error);
      throw error;
    }
  }

  /**
   * 批量添加向量
   * @param {Array} items - 知识库项数组
   * @param {Array} embeddings - 嵌入向量数组
   */
  async addVectorsBatch(items, embeddings) {
    if (!this.isInitialized || !this.collection) {
      // 内存模式
      items.forEach((item, index) => {
        this.memoryCache.set(item.id, { item, embedding: embeddings[index] });
      });
      return items.length;
    }

    try {
      const ids = items.map((item) => item.id);
      const metadatas = items.map((item) => ({
        title: item.title,
        type: item.type,
        created_at: item.created_at,
        updated_at: item.updated_at,
      }));
      const documents = items.map((item) => item.content || item.title);

      await this.collection.add({
        ids,
        embeddings,
        metadatas,
        documents,
      });

      logger.info(`[VectorStore] 批量添加 ${items.length} 个向量`);
      this.emit("vectors-added-batch", items.length);

      return items.length;
    } catch (error) {
      logger.error("[VectorStore] 批量添加向量失败:", error);
      throw error;
    }
  }

  /**
   * 更新向量
   * @param {string} id - 项目ID
   * @param {Array} embedding - 新的嵌入向量
   * @param {Object} metadata - 新的元数据
   */
  async updateVector(id, embedding, metadata) {
    if (!this.isInitialized || !this.collection) {
      // 内存模式
      if (this.memoryCache.has(id)) {
        const cached = this.memoryCache.get(id);
        cached.embedding = embedding;
        if (metadata) {
          Object.assign(cached.item, metadata);
        }
      }
      return;
    }

    try {
      await this.collection.update({
        ids: [id],
        embeddings: [embedding],
        metadatas: metadata ? [metadata] : undefined,
      });

      logger.info(`[VectorStore] 更新向量: ${id}`);
      this.emit("vector-updated", id);

      return true;
    } catch (error) {
      logger.error("[VectorStore] 更新向量失败:", error);
      throw error;
    }
  }

  /**
   * 删除向量
   * @param {string} id - 项目ID
   */
  async deleteVector(id) {
    if (!this.isInitialized || !this.collection) {
      // 内存模式
      this.memoryCache.delete(id);
      return;
    }

    try {
      await this.collection.delete({
        ids: [id],
      });

      logger.info(`[VectorStore] 删除向量: ${id}`);
      this.emit("vector-deleted", id);

      return true;
    } catch (error) {
      logger.error("[VectorStore] 删除向量失败:", error);
      throw error;
    }
  }

  /**
   * 搜索相似向量
   * @param {Array} queryEmbedding - 查询向量
   * @param {number} topK - 返回数量
   * @param {Object} filter - 过滤条件
   */
  async search(queryEmbedding, topK = null, filter = null) {
    topK = topK || this.config.topK;

    if (!this.isInitialized || !this.collection) {
      // 内存模式 - 简单的余弦相似度搜索
      return this.searchMemory(queryEmbedding, topK);
    }

    try {
      const results = await this.collection.query({
        queryEmbeddings: [queryEmbedding],
        nResults: topK,
        where: filter,
      });

      // 格式化结果
      const formatted = [];
      if (results.ids && results.ids[0]) {
        for (let i = 0; i < results.ids[0].length; i++) {
          const similarity = 1 - (results.distances[0][i] || 0); // 转换距离为相似度

          // 过滤低于阈值的结果
          if (similarity >= this.config.similarityThreshold) {
            formatted.push({
              id: results.ids[0][i],
              score: similarity,
              metadata: results.metadatas[0][i],
              document: results.documents[0][i],
            });
          }
        }
      }

      logger.info(`[VectorStore] 搜索返回 ${formatted.length} 个结果`);

      return formatted;
    } catch (error) {
      logger.error("[VectorStore] 搜索失败:", error);
      throw error;
    }
  }

  /**
   * 内存模式搜索
   * @param {Array} queryEmbedding - 查询向量
   * @param {number} topK - 返回数量
   */
  searchMemory(queryEmbedding, topK) {
    const results = [];

    for (const [id, data] of this.memoryCache) {
      const similarity = this.cosineSimilarity(queryEmbedding, data.embedding);

      if (similarity >= this.config.similarityThreshold) {
        results.push({
          id,
          score: similarity,
          metadata: {
            title: data.item.title,
            type: data.item.type,
          },
          document: data.item.content || data.item.title,
        });
      }
    }

    // 按相似度排序
    results.sort((a, b) => b.score - a.score);

    // 返回Top-K
    return results.slice(0, topK);
  }

  /**
   * 计算余弦相似度
   */
  cosineSimilarity(vecA, vecB) {
    if (!vecA || !vecB || vecA.length !== vecB.length) {
      return 0;
    }

    let dotProduct = 0;
    let normA = 0;
    let normB = 0;

    for (let i = 0; i < vecA.length; i++) {
      dotProduct += vecA[i] * vecB[i];
      normA += vecA[i] * vecA[i];
      normB += vecB[i] * vecB[i];
    }

    if (normA === 0 || normB === 0) {
      return 0;
    }

    return dotProduct / (Math.sqrt(normA) * Math.sqrt(normB));
  }

  /**
   * 获取统计信息
   */
  async getStats() {
    if (!this.isInitialized || !this.collection) {
      return {
        mode: "memory",
        count: this.memoryCache.size,
        collectionName: this.config.collectionName,
      };
    }

    try {
      const count = await this.collection.count();

      return {
        mode: "chromadb",
        count,
        collectionName: this.config.collectionName,
        chromaUrl: this.config.chromaUrl,
      };
    } catch (error) {
      logger.error("[VectorStore] 获取统计失败:", error);
      return {
        mode: "error",
        error: error.message,
      };
    }
  }

  /**
   * 清空所有向量
   */
  async clear() {
    if (!this.isInitialized || !this.collection) {
      this.memoryCache.clear();
      return true;
    }

    try {
      // 删除集合并重新创建
      await this.client.deleteCollection({
        name: this.config.collectionName,
      });

      await this.getOrCreateCollection();

      logger.info("[VectorStore] 向量存储已清空");
      this.emit("cleared");

      return true;
    } catch (error) {
      logger.error("[VectorStore] 清空失败:", error);
      throw error;
    }
  }

  /**
   * 重建索引
   * @param {Array} items - 所有知识库项
   * @param {Function} embeddingFn - 嵌入函数
   */
  async rebuildIndex(items, embeddingFn) {
    logger.info(`[VectorStore] 开始重建索引，共 ${items.length} 项`);

    try {
      // 清空现有索引
      await this.clear();

      // 批量生成嵌入并添加
      const batchSize = 10;
      let processed = 0;

      for (let i = 0; i < items.length; i += batchSize) {
        const batch = items.slice(i, i + batchSize);

        // 生成嵌入
        const texts = batch.map(
          (item) => `${item.title}\n${item.content || ""}`,
        );
        const embeddings = await Promise.all(
          texts.map((text) => embeddingFn(text)),
        );

        // 添加到向量存储
        await this.addVectorsBatch(batch, embeddings);

        processed += batch.length;
        this.emit("rebuild-progress", {
          processed,
          total: items.length,
          percentage: Math.round((processed / items.length) * 100),
        });

        logger.info(`[VectorStore] 重建进度: ${processed}/${items.length}`);
      }

      logger.info("[VectorStore] 索引重建完成");
      this.emit("rebuild-complete");

      return true;
    } catch (error) {
      logger.error("[VectorStore] 重建索引失败:", error);
      throw error;
    }
  }

  /**
   * 关闭连接
   */
  async close() {
    logger.info("[VectorStore] 关闭向量存储");

    this.client = null;
    this.collection = null;
    this.isInitialized = false;
    this.memoryCache.clear();

    this.emit("closed");
  }
}

module.exports = VectorStore;
